### Retail Portfolio
__Author__: Toby Limanto

__Objective__: This notebook aims to analyse and visualise stock ownership in IDX per July 2024 based on stock ownership data that is collected from KSEI. https://www.ksei.co.id/archive_download/holding_composition/2022

In [1]:
import pandas as pd

__Data Preprocessing__

Raw data that we get from KSEI can be seen as follows
```
Date|Code|Type|Sec. Num|Price|Local IS|Local CP|Local PF|Local IB|Local ID|Local MF|Local SC|Local FD|Local OT|Total|Foreign IS|Foreign CP|Foreign PF|Foreign IB|Foreign ID|Foreign MF|Foreign SC|Foreign FD|Foreign OT|Total
31-JAN-2022|AALI|EQUITY|1924688333|9900|88378271|15415145|13569998|49700|112448683|31738850|1278471|4138321|5078427|272095866|2356073|8716936|12240314|18338820|1368197|38014142|27063174|886123|9709948|118693727
```

The file contains stock ownership data on a given month. If we are looking at the year 2022, we see 12 files for each month. We will be mostly concerned with the following fields:
- Code: Stock ticker
- Price: Closing stock price at end of month
- Local ID: Number of shares owned by local retail investors

In [6]:
# load the documents
PREFIX = "data-kepemilikan-saham/Balancepos"
date = "20240731"
jul2024 = pd.read_csv(f"{PREFIX}{date}.txt", sep="|")

jul2024 = jul2024[jul2024["Type"] == "EQUITY"]

In [7]:
jul2024.head()

Date  Code    Type      Sec. Num  Price   Local IS     Local CP  \
0  31-JUL-2024  AALI  EQUITY  1.924688e+09   6025   64488127     30698824   
1  31-JUL-2024  ABBA  EQUITY  3.935893e+09     30   71825500   2352371805   
2  31-JUL-2024  ABDA  EQUITY  6.208067e+08   4500          0        40248   
3  31-JUL-2024  ABMM  EQUITY  2.753165e+09   3790    4862900   1481978433   
4  31-JUL-2024  ACES  EQUITY  1.715000e+10    800  724768015  10546239200   

   Local PF  Local IB    Local ID  ...  Foreign IS  Foreign CP  Foreign PF  \
0   6514312     49700   182757347  ...     1626610     6084893    15322570   
1         0         0  1009143877  ...           0       20400           0   
2         0        98      394382  ...   588313138           0           0   
3         0     12000   387388300  ...      550000   825067436     8752441   
4  66622611         0   982577191  ...     6247400   405418474   385441162   

   Foreign IB  Foreign ID  Foreign MF  Foreign SC  Foreign FD  Foreign OT  \
0     9023065     1699364    28119919    24390621      190500     7816149   
1    41020085     1940300           0     1333800           0           0   
2        3600         400           0        4400           0    30499400   
3     4827288      358600    17270334     1681888     1792865    10605409   
4   407431210     2080400  1515175304   140968439    41331600  1224159572   

      Total.1  
0    94273691  
1    44314585  
2   618820938  
3   870906261  
4  4128253561  

[5 rows x 25 columns]

In [12]:
def amount_foreign(row):
    return (row["Foreign IS"] + row["Foreign CP"] + row["Foreign PF"] + row["Foreign IB"] + row["Foreign ID"] + row["Foreign MF"] + row["Foreign SC"] + row["Foreign FD"] + row["Foreign OT"]) * row["Price"] / 1e9

In [17]:
def add_amount(df, f):
    df["Amount (Rp Bio)"] = df.apply(f, axis=1)
    return df.sort_values(by="Amount (Rp Bio)", ascending=False)

In [18]:
foreign_ = add_amount(jul2024, amount_foreign)["Amount (Rp Bio)"].sum()

3322047.0686457027

In [20]:
def amount_local_retail(row):
    return (row["Local ID"]) * row["Price"] / 1e9

In [31]:
local_retail_amount = add_amount(jul2024, amount_local_retail)["Amount (Rp Bio)"].sum()

In [29]:
def amount_local_institution(row):
    return (row["Local IS"] + row["Local CP"] + row["Local PF"] + row["Local IB"] + row["Local MF"] + row["Local SC"] + row["Local FD"] + row["Local OT"]) * row["Price"] / 1e9

In [32]:
local_institution_amount = add_amount(jul2024, amount_local_institution)["Amount (Rp Bio)"].sum()

In [33]:
local_retail_amount/local_institution_amount

0.3680255279154295

### Ratio local retail to local institution

Local Retail: 1102502.4838178921

Local Institution: 2995722.851245368

Ratio: ~37%